vm/src/function_descriptor.h

```diff
@@ -0,0 +1,48 @@
+#ifndef BORGES_FUNCTION_DESCRIPTOR_H
+#define BORGES_FUNCTION_DESCRIPTOR_H
+
+#include "function_traits.h"
+#include "type.h"
+
+namespace Borges {
+
+// Describes the run-time type of a function. In particular, holds the run-time
+// type of its result, the number of its parameters, and the run-time type of
+// each of its parameters.
+struct Function_descriptor {
+    Type result;
+    std::size_t parameters_size;
+    const Type* parameters;
+};
+
+template <typename Parameters>
+struct Describe_parameters;
+
+template <typename... Ts>
+struct Describe_parameters<Parameters<Ts...>> {
+    static constexpr Type value[] = {Type_of<typename std::remove_const<
+        typename std::remove_reference<Ts>::type>::type>::value...};
+};
+
```

Comment:
this is very difficult for me to review, because even though I understand the function of parameter packs, i do not have *any* hands on experience with them. How do we deal with this?